site/src/pages/WorkspacePage/WorkspaceActions/DownloadLogsDialog.tsx

```diff
@@ -28,70 +36,124 @@ type DownloadableFile = {
 
 export const DownloadLogsDialog: FC<DownloadLogsDialogProps> = ({
   workspace,
+  open,
+  onClose,
   download = saveAs,
-  ...dialogProps
 }) => {
   const theme = useTheme();
-  const agents = selectAgents(workspace);
-  const agentLogResults = useQueries({
-    queries: agents.map((a) => ({
-      ...agentLogs(workspace.id, a.id),
-      enabled: dialogProps.open,
-    })),
-  });
+
   const buildLogsQuery = useQuery({
     ...buildLogs(workspace),
-    enabled: dialogProps.open,
+    enabled: open,
   });
-  const downloadableFiles: DownloadableFile[] = useMemo(() => {
-    const files: DownloadableFile[] = [
-      {
-        name: `${workspace.name}-build-logs.txt`,
-        blob: buildLogsQuery.data
-          ? new Blob([buildLogsQuery.data.map((l) => l.output).join("\n")], {
-              type: "text/plain",
-            })
-          : undefined,
-      },
+  const buildLogsFile = useMemo<DownloadableFile>(() => {
+    return {
+      name: `${workspace.name}-build-logs.txt`,
+      blob: buildLogsQuery.data
+        ? new Blob([buildLogsQuery.data.map((l) => l.output).join("\n")], {
+            type: "text/plain",
+          })
+        : undefined,
+    };
+  }, [workspace.name, buildLogsQuery.data]);
+
+  // This is clunky, but we have to memoize in two steps to make sure that we
+  // don't accidentally break the memo cache every render. We can't tuck
+  // everything into a single memo call, because we need to set up React Query
+  // state between processing the agents, but we can't violate rules of hooks by
+  // putting hooks inside of hooks
+  type AgentInfo = Readonly<{
+    agents: readonly WorkspaceAgent[];
+    logOptionsArray: readonly UseQueryOptions<readonly WorkspaceAgentLog[]>[];
+  }>;
+
+  const { agents, logOptionsArray } = useMemo<AgentInfo>(() => {
+    const allAgents = workspace.latest_build.resources.flatMap(
+      (resource) => resource.agents ?? [],
+    );
+
+    // Can't use the "new Set()" trick because we're not dealing with primitives
+    const uniqueAgents = [
+      ...new Map(allAgents.map((agent) => [agent.id, agent])).values(),
     ];
 
+    return {
+      agents: uniqueAgents,
+      logOptionsArray: uniqueAgents.map((agent) => {
```

Comment:
Why do we need to calculate the query options here and not mount it and pass it in the query? I think it is kinda far from what is used and I took some time to link `logOptionsArray` to `agentQueryOptions`

Reply:
Yeah, this was an over-optimization on my part. I basically figured that since the query objects shouldn't change unless the agents do, it was safe to throw them all together